geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/ClearPRMessage.java

```diff
@@ -0,0 +1,405 @@
+    this.resetRecipients();
+    if (recipients != null) {
+      setRecipients(recipients);
+    }
+    this.regionId = region.getPRId();
+    this.processor = replyProcessor;
+    this.processorId = replyProcessor == null ? 0 : replyProcessor.getProcessorId();
+    if (replyProcessor != null && this.isSevereAlertCompatible()) {
+      replyProcessor.enableSevereAlertProcessing();
+    }
+    this.notificationOnly = notifyOnly;
+  }
+
+  @Override
+  public boolean isSevereAlertCompatible() {
+    // allow forced-disconnect processing for all cache op messages
+    return true;
+  }
+
+  public void setPossibleDuplicate(boolean posDup) {
+    this.posDup = posDup;
+  }
+
+  public RegionEventImpl getRegionEvent() {
+    return regionEvent;
+  }
+
+  public ClearResponse send(DistributedMember recipient, PartitionedRegion region)
+      throws ForceReattemptException {
+    Set<InternalDistributedMember> recipients =
+        Collections.singleton((InternalDistributedMember) recipient);
+    ClearResponse clearResponse = new ClearResponse(region.getSystem(), recipients);
+    initMessage(region, recipients, false, clearResponse);
+    setTransactionDistributed(region.getCache().getTxManager().isDistributed());
+    if (logger.isDebugEnabled()) {
+      logger.debug("ClearPRMessage.send: recipient is {}, msg is {}", recipient, this);
+    }
+
+    Set<InternalDistributedMember> failures = region.getDistributionManager().putOutgoing(this);
+    if (failures != null && failures.size() > 0) {
+      throw new ForceReattemptException("Failed sending <" + this + ">");
+    }
+    return clearResponse;
+  }
+
+  @Override
+  public int getDSFID() {
+    return PR_CLEAR_MESSAGE;
+  }
+
+  @Override
+  public void toData(DataOutput out, SerializationContext context) throws IOException {
+    super.toData(out, context);
+    if (bucketId == null) {
+      InternalDataSerializer.writeSignedVL(-1, out);
+    } else {
+      InternalDataSerializer.writeSignedVL(bucketId, out);
+    }
+    DataSerializer.writeObject(this.callbackArg, out);
+    // BR: This may not work properly
+    DataSerializer.writeObject(regionEvent, out);
+  }
+
+  @Override
+  public void fromData(DataInput in, DeserializationContext context)
+      throws IOException, ClassNotFoundException {
+    super.fromData(in, context);
+    this.bucketId = (int) InternalDataSerializer.readSignedVL(in);
+    this.callbackArg = DataSerializer.readObject(in);
+    this.regionEvent = DataSerializer.readObject(in);
+  }
+
+  @Override
+  public EventID getEventID() {
+    return regionEvent.getEventId();
+  }
+
+  /**
+   * This method is called upon receipt and make the desired changes to the PartitionedRegion Note:
+   * It is very important that this message does NOT cause any deadlocks as the sender will wait
+   * indefinitely for the acknowledgement
+   */
+  @Override
+  protected boolean operateOnPartitionedRegion(ClusterDistributionManager distributionManager,
```

Comment:
This method always returns false. Is that intentional? Also, I'm not sure if this method can actually throw any exceptions, since we catch the `ForceReattemptException` from `doLocalClear()` and the other two don't seem to be thrown by anything we're calling here.

Reply:
OperateOnPartitionedRegion() is invoked by the process() method on the message class which is defined in the parent, and the return value is used to indicate to the Process() method whether or not it needs to construct and send a reply. Since we construct a reply ourselves in both return cases in this method, we don't need a case which returns true. 

It also doesn't need to throw any ForceReattemptExceptions since it's bundling those in the reply sent from the case in which those are thrown by the doLocalClear() method.